### PR TITLE
add method to resolve gradient templates

### PR DESCRIPTION
**Pull request by anthrotype**
Some noto-emoji regional flags (e.g. BZ, FK, GT, MX, NI) contain gradients with `xlink:href` pointing to other gradients used as [templates](https://www.w3.org/TR/SVG/pservers.html#PaintServerTemplates). This PR adds support for resolving these references, copying attributes and stops from the template and removing the template if left orphan.

Fixes https://github.com/googlefonts/picosvg/issues/191

### CODE REVIEW

File: src/picosvg/svg.py [hunk 1]
```diff
@@ -944,6 +948,69 @@ def _apply_gradient_translation(self, inplace=False):
...
+        # https://www.w3.org/TR/SVG/pservers.html#PaintServerTemplates
+        if not inplace:
+            svg = SVG(copy.deepcopy(self.svg_root))
+            svg._resolve_gradient_templates(inplace=True)
+            return svg
+
+        href_attr = _xlink_href_attr_name()
+        el_by_id = {el.attrib["id"]: el for el in self.xpath(".//svg:*[@id]")}
+
+        def resolve_gradient_template(gradient):
```
Comment (rsheeter) [inline src/picosvg/svg.py:963]: the nested relatively long function isn't helping with readability much, pop it out?

File: src/picosvg/svg.py [hunk 2]
```diff
@@ -944,6 +948,69 @@ def _apply_gradient_translation(self, inplace=False):
...
+            ref = gradient.attrib[href_attr]
+            if not ref.startswith("#"):
+                raise ValueError(f"Only use #fragment supported, reject {ref}")
+
+            template = el_by_id.get(ref[1:])
```
Comment (rsheeter) [inline src/picosvg/svg.py:968]: suggest ref = ref[1:] once instead of all over

File: src/picosvg/svg.py [hunk 3]
```diff
@@ -944,6 +948,69 @@ def _apply_gradient_translation(self, inplace=False):
...
+                    el = self.resolve_url(shape.fill, "*")
+                except ValueError:  # skip not found
+                    continue
+                if strip_ns(el.tag) not in _GRADIENT_CLASSES:
+                    # unlikely the url target isn't a gradient but I'm not the police
+                    continue
+                used_gradient_ids.add(el.attrib["id"])
+        for grad in self._select_gradients():
+            if grad.attrib.get("id") not in used_gradient_ids:
+                grad.getparent().remove(grad)
```
Comment (rsheeter) [inline src/picosvg/svg.py:1012]: `_safe_remove(grad)`

File: src/picosvg/svg.py (hunk 1 above)
Comment (rsheeter) [reply src/picosvg/svg.py:963]: typing gradient would be nice too

File: src/picosvg/svg.py [hunk 4]
```diff
@@ -944,6 +948,69 @@ def _apply_gradient_translation(self, inplace=False):
...
+
+            # recurse if template references another template
+            if template.attrib.get(href_attr):
+                resolve_gradient_template(template)
+
+            for attr_name in _GRADIENT_ATTRS[strip_ns(gradient.tag)]:
+                if attr_name in template.attrib and attr_name not in gradient.attrib:
+                    gradient.attrib[attr_name] = template.attrib[attr_name]
+
+            if len(gradient) == 0:
```
Comment (rsheeter) [inline src/picosvg/svg.py:987]: len(gradient) threw me; maybe a comment? - `# only copy stops if we don't have our own` or some such

**Review by rsheeter — APPROVED** [review]

File: src/picosvg/svg.py (hunk 1 above)
Comment (anthrotype) [reply src/picosvg/svg.py:963]: The nested function is to be able to recurse

**Pull request merged**